### PR TITLE
Start making patchew more hackable, and some django fixes

### PR DESCRIPTION
**Pull request by bonzini**
First steps towards fixing issue #20.

### CODE REVIEW

File: scripts/developer-setup [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+echo "    . venv/bin/activate"
+echo "    export PYTHONPATH=."
+echo
+echo To run tests, try
+echo
+echo "    ./manage.py test tests"
+echo
+echo To start a web server on http://127.0.0.1:8000, try
+echo
+echo "    gunicorn wsgi"
```
Comment (famz) [inline scripts/developer-setup]: Why did you prefer `gunicorn wsgi` over `./manage.py runserver`?

**Comment by bonzini** [comment]
Because I didn't know anything about Django and copied what the patchew-server script was doing...

File: patchew/settings.py [hunk 2]
```diff
@@ -94,11 +95,14 @@
     DEBUG = os.environ.get("PATCHEW_DEBUG", False)
     ALLOWED_HOSTS = ['*']
 else:
-    if os.environ.get("PATCHEW_TEST"):
+    if os.environ.get("VIRTUAL_ENV"):
+        DEBUG = True
+        DATA_DIR = os.environ.get("VIRTUAL_ENV") + "/data"
+    elif os.environ.get("PATCHEW_TEST"):
+        DEBUG = True
         DATA_DIR = "/tmp/patchew-test-data"
```
Comment (famz) [inline patchew/settings.py]: This breaks non-virtualenv `./manage.py test tests` and `./manage.py runserver` which I am used to do. I'd like to clean up the DEBUG and DATA_DIR initialization code a bit like this:

```patch
commit 263399a7b1c5579154b6f2a09a2ae78258ec4505
Author: Paolo Bonzini <pbonzini@redhat.com>
Date:   Sun Dec 3 07:40:24 2017 +0100

    first step towards using venv
    
    Allow storing the data under the virtualenv subdirectory.
    
    Example:
    
    python3 -m venv ./venv
    . venv/bin/activate
    export PYTHONPATH=.
    pip install -r requirements.txt
    ./manage.py migrate
    ./manage.py collectstatic
    tests/test_testing.py

diff --git a/patchew/settings.py b/patchew/settings.py
index 3ecc7c0..0658ddd 100644
--- a/patchew/settings.py
+++ b/patchew/settings.py
@@ -34,8 +34,9 @@ MODULE_DIR = os.path.join(BASE_DIR, "mods")
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = '@f-l5@70om7o(7rda^oxd$f#60g3jy#&m^p7z@vkf+&$*@%!^o'
 
-DEBUG = True
 ALLOWED_HOSTS = []
+if os.environ.get("PATCHEW_TEST"):
+    ALLOWED_HOSTS = ['testserver']
 
 # Application definition
 
@@ -87,21 +88,30 @@ WSGI_APPLICATION = 'patchew.wsgi.application'
 # Database
 # https://docs.djangoproject.com/en/1.9/ref/settings/#databases
 
-DATA_DIR = os.environ.get("PATCHEW_DATA_DIR")
-if not DATA_DIR:
-    DATA_DIR = os.environ.get("OPENSHIFT_DATA_DIR")
-if DATA_DIR:
-    DEBUG = os.environ.get("PATCHEW_DEBUG", False)
-    ALLOWED_HOSTS = ['*']
-else:
-    if os.environ.get("PATCHEW_TEST"):
-        DATA_DIR = "/tmp/patchew-test-data"
-    elif DEBUG:
-        DATA_DIR = "/var/tmp/patchew-data"
+def env_detect():
+    if "PATCHEW_DATA_DIR" in os.environ:
+        # Docker deployment
+        return False, os.environ.get("PATCHEW_DATA_DIR")
+    elif "OPENSHIFT_DATA_DIR" in os.environ:
+        # OpenShift deployment
+        return False, os.environ.get("OPENSHIFT_DATA_DIR")
+    elif "VIRTUAL_ENV" in os.environ or os.environ.get("PATCHEW_DEBUG", False):
+        # Development setup
+        return True, os.path.join(os.environ.get("VIRTUAL_ENV",
+                                                 "/var/tmp/patchew-dev"),
+                                  "data")
+    elif os.environ.get("PATCHEW_TEST"):
+        # Test environment
+        return True, os.environ.get("PATCHEW_TEST_DATA_DIR")
     else:
-        DATA_DIR = "/data/patchew"
+        raise Exception("Unknown running environment")
+
+DEBUG, DATA_DIR = env_detect()
+if DEBUG:
+    ALLOWED_HOSTS = ["*"]
+
 if not os.path.isdir(DATA_DIR):
-    os.mkdir(DATA_DIR)
+    os.makedirs(DATA_DIR)
 
 BLOB_DIR = os.path.join(DATA_DIR, "blob")
 if not os.path.isdir(BLOB_DIR):
```

I'm testing `scripts/developer-setup` with the revised commit, if it works as expected too, I'll merge this series to master. Look good?

**Comment by famz** [comment]
Applied manually with some tweakings, so closing this PR.

**Pull request closed without merging**